src/steamship/agents/tools/text_generation/json_object_generator.py

```diff
@@ -0,0 +1,170 @@
+
+NEW OBJECT:
+{new_object_prefix}
+"""
+
+DEFAULT_PLURAL_OBJECT_DESCRIPTION = "employees of a company"
+DEFAULT_OBJECT_KEYS = ["Name", "Age", "Gender"]
+DEFAULT_OBJECT_EXAMPLES = [
+    ["Bob", 30, "Male"],
+    ["Susan", 32, "Female"],
+    ["Zhenzhong", 40, "Male"],
+    ["Luis", 32, "Male"],
+    ["Roberta", 35, "Female"],
+    ["Sofia", 30, "Female"],
+]
+DEFAULT_NEW_ROW_PREFIX_FIELDS = []
+
+
+class JsonObjectGeneratorTool(Tool):
+    """
+    Example tool to illustrate generating a new JSON object provided a set of examples.
+
+    This is useful as an example of how to generate a new structured object:
+
+    - A Person (e.g. name, gender, age)
+    - A Proposed Podcast Episode (e.g. title, description, tags)
+
+    The tool takes no input at runtime: it's a true generator parameterized only at initializtion time.
+
+    The tool's parameterization is somewhat CSV-like in construction.
+
+    """
+
+    rewrite_prompt: str = DEFAULT_PROMPT
+    """The prompt used to generate a new JSON object."""
+
+    plural_object_description: str = DEFAULT_PLURAL_OBJECT_DESCRIPTION
+    """Plural description of the object. E.g. 'employees of a company' or 'people' or 'podcast episodes'"""
+
+    object_keys: List[str] = DEFAULT_OBJECT_KEYS
+    """The keys the JSON should have."""
+
+    example_rows: List[List[str]] = DEFAULT_OBJECT_EXAMPLES
+    """List of example object values, aligned to the `object_keys` parameter."""
+
+    new_row_prefix_fields: List[str] = DEFAULT_NEW_ROW_PREFIX_FIELDS
+    """Any fields that should be hard-coded for the new row. These must be grouped as the first set of fields."""
+
+    shuffle_example_rows: bool = True
+    """Whether randomly shuffle example rows to induce a bit of variety even with low LLM temperature."""
+
+    validate_output_as_json: bool = True
+    """Whether to validate that the output is actually JSON."""
+
+    name: str = "JsonObjectTool"
+    human_description: str = "Generates a new JSON object."
+    agent_description: str = "(set at initialization time)"
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.agent_description = (
+            f"Used to generate an instance of the {self.plural_object_description} table. "
+            "Input: Anything. "
+            f"Output A tab-separated row describing a new instance of the {self.plural_object_description} table."
+        )
+
+    def kv_clause(self, key: str, value: str) -> str:
+        """Return an escaped, JSON style key-value clause `"key": "value"`"""
+        value = str(value).replace('"', '\\"')
+        clause = f'"{key}": "{value}"'
+        return clause
+
+    def object_json(self, schema: List[str], values: List[str]):
+        """Render a CSV-style header row and value list into a JSON object."""
+        clauses = []
+        for field, value in zip(schema, values):
+            clauses.append(self.kv_clause(field, value))
+
+        return "{" + ", ".join(clauses) + "}"
```

Comment:
Why are you rolling your own mini-encoder instead of creating a dict of strings and then calling `json.dumps`?

Reply:
I was worried about stable ordering. 

I don't want to just alphabetize it, because that takes away the ability to hard-code some of the fields for the new generation.

E.g. key order has to be [ HARD_CODED_FIELDS ] + [ GENERATED_FIELDS ]

Could def be missing an easier way to do it that I'm not seeing

Reply:
Hard-coded case is, e.g.

Generate me the object `{ story_genre, story_title, story_summary }` but with the `story_genre` fixed to `COMEDY`

So in the case, you need the "Generation row" to be this:

```
{"story_genre": "COMEDY"
```

rather than just

```
{
```

And you also need all of the examples to (I think?) be in the same order too to help the LLM see patterns